### ISSUE TITLE
json: true behaviour confusing

### ISSUE THREAD

**Issue by andenq** [request/request-promise#143]
Ok so I just spent 1 hour trying to figure out why my code wasn't working. I was trying to use the code below, but it just didn't seem to want to work. 

```
options.transform: function (body, response, resolveWithFullResponse) {
    return body.split('').reverse().join('');
};
```

The `body` was already transformed as JSON and I could not for the life of me understand why. _Why why why_ is there a transform function running before **my** transform function, I thought. Well, apparently `options.json` was set to `true` earlier in my code (because my requests were being sent to a json api) and this led to my transform function being rendered completely useless. 

### Workaround
Set `options.json` to `false` and then manually stringify the body before it's sent.

### Suggested solution
Disable the parsing if there is a transform function present. Or just run `options.transform` before parsing the body. _Or_ if this is intended behaviour, please make it more clear that this is going to happen if I switch the `options.json` flag.

**Comment by analog-nico** [comment]
You actually have a good point here. I myself was confused by the `json` option in the beginning.

Fyi, the [`json` option](https://github.com/request/request#requestoptions-callback) is implemented by `request` which is used by `request-promise` to do all the heavy lifting. `request-promise` only adds the promise support and a few extra options like `transform`.

Disabling the parsing – the first part of your suggested solution – would be more confusing in my opinion. Suddenly there is a dependency between two options with a changed behavior that only kicks in if both options are used. But not if either one of them is used. Also, if you are working with a JSON api practical implementations of a transform function operate on the parsed JSON anyway. Why would you want to manipulate the string itself? Or do you?

I certainly agree with your second part of your suggested solution – we need to improve the docs. To date a new user needs gather the pieces at different locations. For one you find the documentation of the `transform` option in `request-promise`'s README and the second piece is the documentation of the `json` option in `request`'s README.

> `json` - sets `body` to JSON representation of value and adds Content-type: application/json header. Additionally, parses the response body as JSON.

I am blind if it comes to that because I know the libraries in and out. What would you say as a fairly new user? What should we improve upon?

- Put the pieces closer together? In the past many users already overlooked that they also need to check the documentation of `request` and thus mostly just missed some important information.
- Or would you say after having read those pieces it is still unclear how the library works in this regard?

**Comment by andenq** [comment]
Hmm ok, my bad I guess hehe :)
Mostly I think the documentation is so good that it's easy to forget that it's just a wrapper. But am I correct in saying that the transform function is an extra function that gets added by `request-promise` ? Because then I can understand why it works as it is.

## Disabling the parsing
That is not exactly how I see it. The `json` option is more or less a transform function on it's own, right? It takes a string and returns JSON. So the way I see it, the `transform` function would override the `json` transform function. But maybe that's just me 😉 

## Improvement
Even after reading the `request` documentation the functionality remains fuzzy to a degree. I understand what the `json` flag does. The fact that cleared it up for me was the knowledge that the `json` functionality comes from `request` and therefore it's behavior cannot be modified by `request-promise`.  It then suddenly became clear that `transform` cannot override anything, because it's not part of the same library as the `json` flag/function.

Maybe just specify that `json` comes from `request` and then put a big(er) warning at the top of the readme for `request-promise` that refers people to `request` for further documentation (just like the one that already exists, just move it up).

**Comment by analog-nico** [comment]
Yes, you got that right. The `json = true` logic is implemented by `request`. So `request-promise` couldn't change it anyway other than manipulating the request options – setting `json = false` etc. And, correct, the `transform` function is something `request-promise` adds on top.

I think disabling the parsing would introduce even more confusion because it would only disable the parsing if the user provides a `transform` function. If no `transform` function is provided the parsing remains enabled. With that we get a conditional link between two parameters. That's just asking for more issues to be opened. ;)

You got great points about how to improve it. The docs have to step up here. As you can see in the commit I just did I tried to improve the README a little bit. However, making the relation between the two libraries very clear is an important aspect I want to keep in mind when working on the upcoming new [documentation website](http://request.js.org). So let's keep this issue open until then.

**Comment by xtxixkx** [comment]
@coroiu Any chance you ever got your workaround working? I've tried multiple things - but none seem to work:

I tried setting `json: false` or just removing transform all together and even tried `transform: body => JSON.parse(JSON.stringify(body))`

But all of them seem to give me an error?
